acquia/mc-cs-plugin-custom-objects
# CO email tokens format fixes + functional test

The main work on this was done on the functional test that is making sure we have the right values in the sent email. I was not able to find a way how to catch the email message directly so I'm checking the email message in the web view which is using the same token replacement code as the email messages.

### Steps to test:
1. Create custom object "Products" with all field types.
2. Create at least 3 custom items of Products. One custom item should have empty values.
3. Create a contact with your email address.
4. Assign both custom objects from 2. to the contact.
5. Create a new template email with content similar to the one I have in the functional test:
```
            Dear George,

            check these values, please:

            ## Default formatting
            Name: {custom-object=products:name | limit=10 | format=default | default=No Value found}
            List: {custom-object=products:country-list-test-field | limit=10 | format=default | default=No Value found}
            Date: {custom-object=products:date-test-field | limit=10 | format=default | default=No Value found}
            Datetime: {custom-object=products:datetime-test-field | limit=10 | format=default | default=No Value found}
            Email: {custom-object=products:email-test-field | limit=10 | format=default | default=No Value found}
            Hidden: {custom-object=products:hidden-test-field | limit=10 | format=default | default=No Value found}
            Number: {custom-object=products:number-test-field | limit=10 | format=default | default=No Value found}
            Phone: {custom-object=products:phone-number-test-field | limit=10 | format=default | default=No Value found}
            Select: {custom-object=products:select-test-field | limit=10 | format=default | default=No Value found}
            Multiselect: {custom-object=products:multiselect-test-field | limit=10 | format=default | default=No Value found}
            Text: {custom-object=products:text-test-field | limit=10 | format=default | default=No Value found}
            Textarea: {custom-object=products:textarea-test-field | limit=10 | format=default | default=No Value found}
            Url: {custom-object=products:url-test-field | limit=10 | format=default | default=No Value found}

            ## Various formatting and limit
            Name: {custom-object=products:name | limit=10 | format=or-list | default=No Value found}
            List: {custom-object=products:country-list-test-field | limit=10 | format=and-list | default=No Value found}
            Date: {custom-object=products:date-test-field | limit=10 | format=bullet-list | default=No Value found}
            Datetime: {custom-object=products:datetime-test-field | limit=10 | format=ordered-list | default=No Value found}
            Email: {custom-object=products:email-test-field | limit=1 | format=default | default=No Value found}
            Hidden: {custom-object=products:hidden-test-field | limit=10 | format=bullet-list | default=No Value found}
            Number: {custom-object=products:number-test-field | limit=10 | format=and-list | default=No Value found}
            Phone: {custom-object=products:phone-number-test-field | limit=10 | format=default | default=No Value found}
            Select: {custom-object=products:select-test-field | limit=10 | format=ordered-list | default=No Value found}
            Multiselect: {custom-object=products:multiselect-test-field | limit=10 | format=ordered-list | default=No Value found}
            Text: {custom-object=products:text-test-field | limit=10 | format=bullet-list | default=No Value found}
            Textarea: {custom-object=products:textarea-test-field | limit=10 | format=or-list | default=No Value found}
            Url: {custom-object=products:url-test-field | limit=10 | format=ordered-list | default=No Value found}
```
Notice the different formats for different field types.
6. Send a direct email to the contact from the contact detail view. Select your email before you send it.
7. Check the values in the email make sense. There should not be any of ` , value1, value2` or empty or duplicated values in one token.
